repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeRoutingServlet.scala

```diff
@@ -0,0 +1,362 @@
+                // and is matched against the context roots of other web applications hosted on this container."
+                Option(getServletContext.getContext(locationUrl.getPath)).foreach { targetContext =>
+                  // Capture this for Location header processing
+                  val locationUrlPath = locationUrl.getPath
+                  val targetContextPath = targetContext.getContextPath
+                  val dispatchPath = if (locationUrlPath.startsWith(targetContextPath)) locationUrlPath.substring(targetContextPath.length) else locationUrlPath
+                  Option(targetContext.getRequestDispatcher(dispatchPath)) match {
+                    case Some(dispatcher) =>
+                      doDispatch(
+                        servletRequest,
+                        servletResponse,
+                        locationUrl,
+                        dispatchPath,
+                        targetContext,
+                        routeDest,
+                        dispatcher,
+                        destination)
+                    case _ =>
+                      logger.warn(s"No Request Dispatcher found in Servlet Context for $dispatchPath; not forwarding")
+                  }
+                }
+              case _ =>
+                logger.warn("No route-able node found; not forwarding")
+            }
+          case _ =>
+            logger.warn("No valid local host defined; not forwarding")
+        }
+      case id =>
+        val localId = localCluster.get().map(_.getId) match {
+          case Some(locId) => locId
+          case None => ""
+        }
+        logger.warn("Invalid routing destination specified: {} for cluster: {} ", id, localId)
+        servletResponse.setStatus(HttpServletResponse.SC_NOT_FOUND)
+    }
+  }
+
+  def getDestinationLocation(servletRequest: HttpServletRequestWrapper, destination: Destination, localhost: Node, uri: String, request: HttpServletRequest): Option[URL] = {
+    val http = if (localhost.getHttpPort > 0) List(new Port("http", localhost.getHttpPort)) else List.empty[Port]
+    val https = if (localhost.getHttpsPort > 0) List(new Port("https", localhost.getHttpsPort)) else List.empty[Port]
+    val localPorts = http ++ https
+
+    destination match {
+      case endpoint: DestinationEndpoint =>
+        Some(buildEndpoint(localhost, localPorts, endpoint, uri, servletRequest))
+      case cluster: DestinationCluster =>
+        // TODO: Remove this when routing to a cluster is removed.
+        domains.get().map(_.getDomain(cluster.getCluster.getId)) match {
+          case Some(clusterWrapper) =>
+            val routableNode = clusterWrapper.getNextNode
+            val port = if (HttpsProtocol.equalsIgnoreCase(cluster.getProtocol)) routableNode.getHttpsPort else routableNode.getHttpPort
+            Some(new URL(cluster.getProtocol, routableNode.getHostname, port, cluster.getRootPath + uri))
```

Comment:
Hopefully this goes way, but should `StringUriUtilities.concatUris(cluster.getRootPath, uri)` be used here?

Reply:
Sure.
Updated.